src/main/java/com/google/devtools/build/lib/analysis/actions/LocalTemplateExpansionStrategy.java

```diff
@@ -44,18 +44,43 @@ public ImmutableList<SpawnResult> expandTemplate(
       TemplateExpansionContext.TemplateMetadata templateMetadata)
       throws InterruptedException, ExecException {
     try {
+      // If writeOutputToFile may retain the writer, make sure that it doesn't capture the
+      // expanded template string. Since expansion may throw and the writer must not, expand the
+      // template once before calling writeOutputToFile. It is assumed to be deterministic, so if
+      // it doesn't throw once, it won't throw again.
       final String expandedTemplate =
           getExpandedTemplateUnsafe(
               templateMetadata.template(), templateMetadata.substitutions(), ctx.getPathResolver());
-      DeterministicWriter deterministicWriter =
-          out -> out.write(expandedTemplate.getBytes(ISO_8859_1));
-      return ctx.getContext(FileWriteActionContext.class)
-          .writeOutputToFile(
-              action,
-              ctx,
-              deterministicWriter,
-              templateMetadata.makeExecutable(),
-              /* isRemotable= */ true);
+      FileWriteActionContext fileWriteActionContext = ctx.getContext(FileWriteActionContext.class);
+      DeterministicWriter deterministicWriter;
+      if (fileWriteActionContext.mayRetainWriter()) {
+        ArtifactPathResolver pathResolver = ctx.getPathResolver();
+        deterministicWriter =
+            out -> {
+              try {
+                out.write(
+                    StringUnsafe.getInternalStringBytes(
+                        getExpandedTemplateUnsafe(
+                            templateMetadata.template(),
+                            templateMetadata.substitutions(),
+                            pathResolver)));
+              } catch (EvalException e) {
+                throw new IllegalStateException(
+                    "Template expansion is not deterministic, first succeeded and then failed with: "
+                        + e.getMessage(),
+                    e);
```

Comment:
It looks like the `getExpandedTemplateUnsafe` implementation doesn't actually throw `EvalException` (and there are no google3 overrides). Can we just amend the method signature?